lib/mongo/ObserveMultiplex.js

```diff
@@ -0,0 +1,289 @@
+        if (!self._ready() && !options.fromQueryError)
+            throw Error('surprising _stop: not ready');
+
+        // Call stop callback (which kills the underlying process which sends us
+        // callbacks and removes us from the connection's dictionary).
+        self._onStop();
+        Package['facts-base'] &&
+            Package['facts-base'].Facts.incrementServerFact(
+                'mongo-livedata',
+                'observe-multiplexers',
+                -1
+            );
+
+        // Cause future addHandleAndSendInitialAdds calls to throw (but the onStop
+        // callback should make our connection forget about us).
+        self._handles = null;
+    },
+
+    // Allows all addHandleAndSendInitialAdds calls to return, once all preceding
+    // adds have been processed. Does not block.
+    ready: function() {
+        var self = this;
+        self._queue.queueTask(function() {
+            if (self._ready())
+                throw Error("can't make ObserveMultiplex ready twice!");
+            self._readyFuture.return();
+        });
+    },
+
+    // If trying to execute the query results in an error, call this. This is
+    // intended for permanent errors, not transient network errors that could be
+    // fixed. It should only be called before ready(), because if you called ready
+    // that meant that you managed to run the query once. It will stop this
+    // ObserveMultiplex and cause addHandleAndSendInitialAdds calls (and thus
+    // observeChanges calls) to throw the error.
+    queryError: function(err) {
+        var self = this;
+        self._queue.runTask(function() {
+            if (self._ready())
+                throw Error("can't claim query has an error after it worked!");
+            self._stop({ fromQueryError: true });
+            self._readyFuture.throw(err);
+        });
+    },
+
+    // Calls "cb" once the effects of all "ready", "addHandleAndSendInitialAdds"
+    // and observe callbacks which came before this call have been propagated to
+    // all handles. "ready" must have already been called on this multiplexer.
+    onFlush: function(cb) {
+        var self = this;
+        self._queue.queueTask(function() {
+            if (!self._ready())
+                throw Error(
+                    'only call onFlush on a multiplexer that will be ready'
+                );
+            cb();
+        });
+    },
+    callbackNames: function() {
+        var self = this;
+        if (self._ordered)
+            return ['addedBefore', 'changed', 'movedBefore', 'removed'];
+        else return ['added', 'changed', 'removed'];
+    },
+    _ready: function() {
+        return this._readyFuture.isResolved();
+    },
+    _applyCallback: function(callbackName, args) {
+        var self = this;
+
+        const isOptimistic = !!OptimisticInvocation.get();
+        // TODO Add a debug message here
+        const runType = isOptimistic ? 'runTask' : 'queueTask';
```

Comment:
@theodorDiaconu instead of forcing optimistic invocations to be synchronous, could we try -

given a helper function like:
```js
function maybeBeginWrite(){
  const fence = !!OptimisticInvocation.get() && DDPServer._CurrentWriteFence.get();
  if (fence) {
    return fence.beginWrite();
  } else {
    return {committed: function () {}};
  }
}
```

We can then write:

```
const fence = maybeBeginWrite();

self._queue.queueTask(function(){
   try {
      /* ... */
   } finally {
      fence.committed();
  }
});
```